File: app/src/main/java/com/kickstarter/viewmodels/ProjectNotificationSettingsViewModel.java
package com.kickstarter.viewmodels;

import com.kickstarter.libs.ActivityViewModel;
import com.kickstarter.libs.Environment;
import com.kickstarter.models.ProjectNotification;
import com.kickstarter.services.ApiClientType;
import com.kickstarter.ui.activities.ProjectNotificationSettingsActivity;

import java.util.List;

import androidx.annotation.NonNull;
import rx.Notification;
import rx.Observable;
import rx.subjects.BehaviorSubject;

import static com.kickstarter.libs.rx.transformers.Transformers.errors;
import static com.kickstarter.libs.rx.transformers.Transformers.values;

public interface ProjectNotificationSettingsViewModel {

  interface Outputs {
    Observable<List<ProjectNotification>> projectNotifications();

    Observable<Void> unableToFetchProjectNotificationsError();
  }

  final class ViewModel extends ActivityViewModel<ProjectNotificationSettingsActivity> implements Outputs {

    private final ApiClientType client;

    public ViewModel(final @NonNull Environment environment) {
      super(environment);

      this.client = environment.apiClient();

      final Observable<Notification<List<ProjectNotification>>> projectNotificationsNotification =
        this.client
          .fetchProjectNotifications()
          .materialize();

      projectNotificationsNotification
        .compose(values())
        .compose(bindToLifecycle())
        .subscribe(this.projectNotifications::onNext);

      projectNotificationsNotification
        .compose(errors())
        .compose(bindToLifecycle())
        .subscribe(e -> this.unableToFetchProjectNotificationsError.onNext(null));
    }

    private final BehaviorSubject<List<ProjectNotification>> projectNotifications = BehaviorSubject.create();
    private final BehaviorSubject<Void> unableToFetchProjectNotificationsError = BehaviorSubject.create();

    public final Outputs outputs = this;

    @Override public @NonNull Observable<List<ProjectNotification>> projectNotifications() {
      return this.projectNotifications;
    }
    @Override public @NonNull Observable<Void> unableToFetchProjectNotificationsError() {
      return this.unableToFetchProjectNotificationsError;
    }
  }
}

File: app/src/main/java/com/kickstarter/viewmodels/ProjectNotificationSettingsViewModel.kt
package com.kickstarter.viewmodels

import androidx.lifecycle.ViewModelProvider
import com.kickstarter.libs.Environment
import com.kickstarter.libs.rx.transformers.Transformers
import com.kickstarter.libs.utils.extensions.addToDisposable
import com.kickstarter.libs.utils.extensions.isNotNull
import com.kickstarter.models.ProjectNotification
import com.kickstarter.services.ApiClientTypeV2
import io.reactivex.Observable
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.subjects.BehaviorSubject

interface ProjectNotificationSettingsViewModel {
    interface Outputs {
        fun projectNotifications(): Observable<List<ProjectNotification>>
        fun unableToFetchProjectNotificationsError(): Observable<Unit>
    }

    class ProjectNotificationSettingsViewModel(environment: Environment) : androidx.lifecycle.ViewModel(), Outputs {
        private val client: ApiClientTypeV2 = requireNotNull(environment.apiClientV2())
        private val projectNotifications = BehaviorSubject.create<List<ProjectNotification>>()
        private val unableToFetchProjectNotificationsError = BehaviorSubject.create<Unit>()
        private val disposables = CompositeDisposable()

        val outputs: Outputs = this

        override fun projectNotifications(): Observable<List<ProjectNotification>> = projectNotifications

        override fun unableToFetchProjectNotificationsError(): Observable<Unit> = unableToFetchProjectNotificationsError

        init {
            val projectNotificationsNotification = client
                .fetchProjectNotifications()
                .materialize()

            projectNotificationsNotification
                .compose(Transformers.valuesV2())
                .filter { it.isNotNull() }
                .subscribe {
                    projectNotifications.onNext(it)
                }
                .addToDisposable(disposables)

            projectNotificationsNotification
                .compose(Transformers.errorsV2())
                .subscribe { unableToFetchProjectNotificationsError.onNext(Unit) }
                .addToDisposable(disposables)
        }

        override fun onCleared() {
            super.onCleared()
            disposables.clear()
        }
    }

    class Factory(private val environment: Environment) : ViewModelProvider.Factory {
        override fun <T : androidx.lifecycle.ViewModel> create(modelClass: Class<T>): T {
            return ProjectNotificationSettingsViewModel(environment) as T
        }
    }
}
